scripts/installer_test.go

```diff
@@ -0,0 +1,339 @@
+	}
+	install(t, ctx, *msiPathFlag, installDir, args...)
+	defer requireUninstall(t, ctx)
+
+	expectedEmail := fmt.Sprintf("operator.email: %s", email)
+	expectedWallet := fmt.Sprintf("operator.wallet: \"%s\"", walletAddr)
+	expectedAddr := fmt.Sprintf("contact.external-address: %s", publicAddr)
+
+	configStr := readConfigLines(t, ctx, configPath)
+
+	require.Contains(t, configStr, expectedEmail)
+	require.Contains(t, configStr, expectedWallet)
+	require.Contains(t, configStr, expectedAddr)
+}
+
+func TestUpgrade_Config(t *testing.T) {
+	if *msiPathFlag == "" {
+		t.Fatal("no msi passed, use `go test ... -args -msi <msi path>`")
+	}
+
+	ctx := testcontext.New(t)
+	defer ctx.Cleanup()
+
+	installDir := ctx.Dir("install")
+	configPath := ctx.File("install", "config.yaml")
+
+	walletAddr := "0x0000000000000000000000000000000000000000"
+	email := "user@mail.test"
+	publicAddr := "127.0.0.1:10000"
+
+	// install old release
+	args := []string{
+		fmt.Sprintf("STORJ_WALLET=%s", walletAddr),
+		fmt.Sprintf("STORJ_EMAIL=%s", email),
+		fmt.Sprintf("STORJ_PUBLIC_ADDRESS=%s", publicAddr),
+	}
+	install(t, ctx, releaseMSIPath, installDir, args...)
+
+	// upgrade using test msi
+	install(t, ctx, *msiPathFlag, installDir)
+	defer requireUninstall(t, ctx)
+
+	expectedEmail := fmt.Sprintf("operator.email: %s", email)
+	expectedWallet := fmt.Sprintf("operator.wallet: \"%s\"", walletAddr)
+	expectedAddr := fmt.Sprintf("contact.external-address: %s", publicAddr)
+
+	configStr := readConfigLines(t, ctx, configPath)
+
+	require.Contains(t, configStr, expectedEmail)
+	require.Contains(t, configStr, expectedWallet)
+	require.Contains(t, configStr, expectedAddr)
+}
+
+func install(t *testing.T, ctx *testcontext.Context, msiPath, installDir string, args ...string) {
+	t.Logf("installing from %s\n", msiPath)
+	logPath := ctx.File("log", "install.log")
+	args = append(append([]string{
+		"/i", msiPath,
+		"/log", logPath,
+	}, append(msiBaseArgs, "INSTALLFOLDER="+installDir)...), args...)
+
+	installOut, err := exec.Command("msiexec", args...).CombinedOutput()
+	if !assert.NoError(t, err) {
+		installLogData, err := ioutil.ReadFile(logPath)
+		if assert.NoError(t, err) {
+			t.Logf("MSIExec log:\n============================\n%s", string(installLogData))
+		}
+		t.Logf("MSIExec output:\n============================\n%s", string(installOut))
+		t.Fatal()
+	}
+}
+
+func requireUninstall(t *testing.T, ctx *testcontext.Context) {
+	logPath := ctx.File("log", "uninstall.log")
+	uninstallOut, err := uninstall(logPath)
+	if !assert.NoError(t, err) {
+		uninstallLogData, err := ioutil.ReadFile(logPath)
+		if !assert.NoError(t, err) {
+			t.Logf("MSIExec log:\n============================\n%s\n", string(uninstallLogData))
+		}
+		t.Fatalf("MSIExec output:\n============================\n%s", string(uninstallOut))
+	}
+}
+
+func uninstall(logPath string) ([]byte, error) {
+	if err := stopServices("storagenode", "storagenode-updater"); err != nil {
+		return []byte{}, err
+	}
```

Comment:
Why are we stopping the services here? They are supposed to be stopped by the uninstaller.

Reply:
If we don't wait for them to be stopped `ctx.Cleanup()` can cause failures. My thinking was that it's easier to stop them manually than try and open a service that may or may not exist after uninstall and query it until it is stopped.